### ISSUE TITLE
More consistency with array-methods

### ISSUE THREAD

**Issue by petered** [numpy/numpy#7452]
It's sometimes hard to tell in numpy whether a given numpy function will also be a method of an array.

e.g. You can go `arr.max()` or `np.max(arr)`, but you can's go `arr.abs()` (you need to go `np.abs(arr)`.

Same goes for most functions: `median`, `sign`, `exp`, etc.  It would be nice, because I find it much more readable to look at: 

```
mean_rmse = ((out-targ)**2).mean(axis=1).sqrt().mean(axis=0)
```

Than:

```
mean_rmse = np.mean(np.sqrt(np.mean((out-targ)**2, axis =1)), axis=0)
```

**Comment by njsmith** [comment]
Yeah. Unfortunately, there's not much we can do about this :-/

We could make it more consistent by removing methods from ndarray and sticking to just functions. But this is impossible because it would break everyone's existing code that uses methods.

Or, we could move all functions to also be methods. But this is impossible because new users and packages are constantly defining new functions. Plus continuing to multiply these duplicate methods violates "there should be one obvious way to do it".

If we could go back in time then I'd probably argue for not having these methods on ndarray at all, and using functions exclusively. In python, methods are a closed set of operations that have to be specified at class creation time, can't be extended, and can be overridden by subclasses; functions are an open set of operations that can be added to at any time, namespaced, and can't be overridden by subclasses (without some special work anyway). In the ndarray case, it's turned out that method subclassing is ineffective, and that there are many many operations with new ones added all the time, that these can't be written in terms of a small finite set of core operations, and thus that we need to solve the problem of overriding functions for new types anyway. So this all argues for using functions exclusively (sorry if you find it less readable). Too bad we don't have a time machine :-)

**Comment by petered** [comment]
Of all of these options, I would think that the least bad is to make it so that all numpy functions with an array as their first argument are also methods of array.  

+ You have the option to use the more readable "method" syntax
+ It's predictable: If a function `np.func(arr, ...)` exists, then `arr.func(...)` also exists.
+ It's backwards-compatible, no time-machine required
- Somewhat violates the "one way to do things" rule
- If using the "method" syntax, you can no longer treat lists and 1d-arrays interchangeably.
- You bring up that it's awkward to have an ever-expanding method list, with methods just pointing to functions in various different modules (e.g. `def abs(self): return np.abs(self)`).  I don't fully get what the problem is, but I guess it's because Array methods would call functions which themselves reference the Array class, so you get circular imports?  This could be solved by Dynamically binding methods to the Array class when importing the function modules, which, although maybe strange, seems like more of an internal implementation detail of numpy that nobody has to look at, or do I not get it?

**Comment by charris** [comment]
>If we could go back in time then I'd probably argue for not having these methods on ndarray at all, and using functions exclusively.

That was actually how Numeric did things, it was a conscious, thought out decision on the part of the developers. There were, IIRC, two, maybe three methods attached to the Numeric array type. I think is was a good decision, and our best bet to return to that would be to layer ndarray on top of a simpler type, something more like a pure buffer type.

I'm not sure when the added methods first made an appearance, but I suspect it was in Numarray and that numpy inherited them in the forced marriage between Numeric and Numarray that birthed Numpy and then added a few of its own. I might be wrong about that however. @teoliphant would know.

**Comment by njsmith** [comment]
> our best bet to return to that would be to layer ndarray on top of a simpler type, something more like a pure buffer type.

That's a really interesting idea. Now you've got me trying to think how feasible it would be to pull it off :-). It wouldn't have to be an abi break because places that expect ndarrays could get views of this simpler thing. The big challenges that come to mind are (a) allocating two objects every time we create an array would be very difficult to do without unacceptable speed regressions. (b) it's not clear how anyone could actually take advantage of the cut down api if `np.array`, `np.zeros`, etc. all have to continue returning old-style ndarrays.

Also I guess there is code out there that assumes that newly allocated arrays always own their data.

I guess our best bet for pulling anything like this off would be to continue with the plan of making duck arrays more and more viable, and prototype it as a duck-array library. If we wanted to be extra cheeky we could call the slimmed down numpy `np` (so you'd just write `import np`) ;-). As long as the two libraries agreed on array protocols and dtypes and ufuncs then it might be possible...

**Comment by petered** [comment]
Bump!  I'd really like to have the method syntax enabled for the next version of numpy.  It's annoying never knowing when I can use `arr.func(axis=1)` and when I'm restricted to `np.func(arr, axis=1)`.  What's the verdict on this?

**Comment by njsmith** [comment]
@petered: I'm sorry, but I don't think we're going to be adding hundreds of redundant methods to `ndarray`. If you're unsure what to use, I suggest using `np.func(arr, axis=1)`, since that always works :-)

**Comment by petered** [comment]
Daaaaaaaamn.  Really, what's the main reason why not?  The current version is so inconsistent!

**Comment by njsmith** [comment]
I'm a bit confused at why this is a surprise, given that everyone else commenting in this issue has been talking about why it's a bad idea and whether there's any way we can *remove* methods from ndarray.

**Comment by petered** [comment]
Well, we've established that we can't remove them for backwards-compatibility reasons.  So it seems like the next-best thing (or the best thing, as far as code-readability is concerned) is to add them all so the API is consistent.  I have yet to see a good reason why this is impossible.

You say:

> Or, we could move all functions to also be methods. But this is impossible because new users and packages are constantly defining new functions. Plus continuing to multiply these duplicate methods violates "there should be one obvious way to do it".

- I don't see what makes it impossible to implement - it seems like it could be done by dynamically binding methods to the array class in numpy/__init__.py.  Is the problem that it would slow down the import too much?
- The "there should be one obvious way to do it" has already been broken from the start.

**Comment by petered** [comment]
Another argument for method syntax (and then I'll shut up...)

Nowadays libraries like theano and tensorflow are use numpy syntax.

I often end up re-writing functions that I implement in numpy, in theano.  The only difference is that I replace `np.function_name(x, ...)` (after import numpy as np) with `tt.function_name(x, ...)` (after import theano.tensor as tt).

This problem would go away if we used the method syntax.  (We'd just have `x.function_name(...)` in both cases.

**Comment by eric-wieser** [comment]
A better form of that argument for method syntax - polymorphism. If I write a function that works on numpy arrays, I want it to work for my custom subclass of numpy arrays as well.

What if my subclass is `np.ma.MaskedArray`, and wants `sum` to be overidden for it?

Having to type the module the free function belongs to is not only increased typing, but it totally cuts out the option for dropping in derived types.

One alternative is to apply the `@functools.singledispatch` decorator to all the free functions, so that other modules can override those operations that way instead. This probably has a greater performance cost though, and requires python 3.4+

**Comment by sg-s** [comment]
@njsmith you said "here should be one obvious way to do it" 

I present the following ways in which you can do things with numpy arrays:

| What I want | `foo(thing)` | `np.foo(thing)` | `thing.foo` | `thing.foo()` |
| ------------ | ---------- | --------  | ----------- | ---- |
| shape        | ❌ |  ✅      |  ✅   | ❌ | 
| mean        |  ❌        |  ✅ | ❌  | ✅ | 
| diff        |  ❌ | ✅  | ❌ | ❌| 
| dtype |       ❌         |  ❌              |  ✅       |  ❌ |

To my eye it doesn't look like there's one way to do it, and it's not obvious, and it's not consistent.

**Comment by rgommers** [comment]
The active discussion here is old, and it's not very actionable without a lot of breakage. Cleaning up the NumPy API remains an eternal work-in-progress, so I don't think this issue needs to be kept open.

I'll also note that the [array API standard](https://data-apis.org/array-api/latest/) uses _only functions_ and no methods - for the reasons discussed in this issue. Over time, we can make that API more prominent, while de-emphasizing the myriad other things NumPy has that are kept for backwards compatibility reasons.

Thanks all.

**Comment by petered** [comment]
Thanks Ralf, the array API standard looks like a great initiative.  Though I still after all these years would strongly prefer a method-based syntax, both for readability (avoiding the bracket-matching nightmare discussed at top of post) and ability to re-use code like 
```
def sigmoid(x): 
    return 1/(1+(-x).exp())
```
And have it be automatically valid in numpy, tensorflow, pytorch, etc.
